src/2LPoleFunctions.f90

```diff
@@ -0,0 +1,1801 @@
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! MASS FUNCTIONS
+
+
+real(dp) Function WfSSSS (p2,x_in,y_in,z_in,u_in,Q2)
+Implicit None
+real(dp), intent(in):: p2,x_in,y_in,z_in,u_in,q2
+real(dp) :: res,x,y,z,u
+
+x=x_in
+y=y_in
+z=z_in
+u=u_in
+
+
+If(smalldp(x,y)) then
+   res=real(-1._dp*IIp(x,z,u,q2),dp)
+else
+   res=Real((II(x,z,u,q2) - II(y,z,u,q2))/(y-x),dp)
+end if
+
+
+WfSSSS=res
+
+end function WfSSSS
+
+
+real(dp) Function XfSSS(p2,x,y,z,Q2)
+
+Implicit None
+real(dp), intent(in):: p2,x,y,z,q2
+real(dp) :: res
+
+if (smallc(z/q2)) then
+   res = 0._dp
+else
+  res = real(-J0(z,Q2)*BB(x,y,q2),dp)
+end if
+
+XfSSS = res
+end function XfSSS
+
+
+real(dp) Function YfSSSS(p2,x,y,z,u,Q2)
+real(dp), intent(in):: p2,x,y,z,u,q2
+
+real(dp) :: res
+
+
+if(smallc(u/q2)) then
+   res = 0._dp
+else
+   res = real(J0(u,q2)*CC(x,y,z,q2),dp)
+end if
+
+YfSSSS=res
+
+
+end Function YfSSSS
+
+
+real(dp) Function ZfSSSS(p2,x,y,z,u,Q2)
+real(dp), intent(in):: p2,x,y,z,u,q2
+
+real(dp) :: res
+
+
+res = real(BB(x,y,q2)*BB(z,u,q2),dp)
+ZfSSSS=res
+end function ZfSSSS
+
+real(dp) Function SfSSS(p2,x,y,z,Q2)
+real(dp), intent(in):: p2,x,y,z,q2
+
+real(dp) :: res
+
+
+res = -real(II(x,y,z,q2),dp)
+
+SfSSS=res
+end function SfSSS
+
+real(dp) Function UfSSSS(p2,x,y,z,u,Q2)
+real(dp), intent(in):: p2,x,y,z,u,q2
+
+real(dp) :: res
+
+res = UU(x,y,z,u,q2)
+
+UfSSSS=res
+end function UfSSSS
+
+real(dp) Function VfSSSSS(p2,x_in,y_in,z_in,u_in,v_in,Q2)
+real(dp), intent(in):: p2,x_in,y_in,z_in,u_in,v_in,q2
+
+real(dp) :: res,u,v,x,y,z
+
+if(abs(v_in) .gt. abs(u_in)) then
+   v = v_in
+   u = u_in
+else
+   v = u_in
+   u = v_in
+end if
+
+x=x_in
+y=y_in
+z=z_in
+
+if(smalldp(x,y)) then
+       res=real(-VV0(z,x,u,v,q2),dp)
+else
+   if(smalldp(y,z)) then
+      res=real(-VV0(x,y,u,v,q2),dp)
+   else
+      res=real((UU(x,y,u,v,q2)-UU(x,z,u,v,q2))/(y-z),dp)
+   end if
+end if
+
+
+
+VfSSSSS=res
+end function VfSSSSS
+
+real(dp) Function MfSSSSS(p2,x,y,z,u,v,Q2)
+real(dp), intent(in):: p2,x,y,z,u,v,q2
+
+real(dp) :: res
+
+
+
+res=-real(MM(x,y,z,u,v,q2),dp)
+
+MfSSSSS=res
+end function MfSSSSS
+
+!!!!!!!!!!!!!!!!! SCALARS AND VECTORS
+
+
+real(dp) Function WfSSSV(p2,x,Q2)
+real(dp), intent(in):: p2,x,q2
+
+real(dp) :: res
+
+!res=real(3*II(0._dp,x,x,q2)-J0(x,q2)+2._dp*x,dp)
+if(smallc(x/q2)) then
+   res = 0._dp
+else
+   res=real(3._dp*x*(-5._dp+4._dp*log(x/q2)-(log(x/q2))**2)-J0(x,q2)+2._dp*x,dp)
+end if
+
+WfSSSV=res
+
+end function WfSSSV
+
+real(dp) Function MfSSSSV(p2,x_in,y_in,Q2)
+real(dp), intent(in):: p2,x_in,y_in,q2
+
+real(dp) :: res,x,y
+
+if(abs(y_in) .gt. abs(x_in)) then
+   y = y_in
+   x= x_in
+else
+   y = x_in
+   x =y_in
+end if
+
+if(smalldp(x,y)) then
+   call saveme(q2,y) !!!! y > x so just maybe ..
+   res = real( -1 + 5*log(y/q2) -3 *log(y/q2)**2,dp)
+else
+   if(smallc(x/q2)) then
+     !!! NB the case that they are both small is dealt with above ...
+      res=real(12 + (11*(y*log(y/q2))-3*(y*(log(y/q2))**2))/(y),dp)
```

Comment:
This doesn't have to be `dp`?

Reply:
I think you are right, since the addition and multiplication of a `dp` float should give again a `dp` float in FORTRAN.